### PR TITLE
General cleanup in RemotingOkHttpCall

### PR DESCRIPTION
**Pull request by carterkozak**
## After this PR
==COMMIT_MSG==
General safety and cleanup in RemotingOkHttpCall
==COMMIT_MSG==

### CODE REVIEW

File: okhttp-clients/src/main/java/com/palantir/conjure/java/okhttp/RemotingOkHttpCall.java [hunk 1]
```diff
@@ -118,8 +120,12 @@ public void onFailure(Call _call, IOException exception) {
             }
 
             @Override
-            public void onResponse(Call _call, Response response) {
-                future.set(response);
+            public void onResponse(Call call, Response response) {
+                if (!future.set(response)) {
```
Comment (carterkozak) [inline okhttp-clients/src/main/java/com/palantir/conjure/java/okhttp/RemotingOkHttpCall.java:131]: Unlikely this is ever hit, but if it is, we want to know about it.

File: okhttp-clients/src/main/java/com/palantir/conjure/java/okhttp/RemotingOkHttpCall.java [hunk 2]
```diff
@@ -274,16 +290,19 @@ public void onResponse(Call call, Response response) throws IOException {
                     return;
                 }
 
+                ResponseBody maybeResponseBody = response.body();
                 // Buffer the response into a byte[] so that multiple handler can safely consume the body.
                 // This consumes and closes the original response body.
                 Supplier<Response> errorResponseSupplier;
                 try {
-                    byte[] body = response.body().bytes();
+                    byte[] body = maybeResponseBody == null ? null : maybeResponseBody.bytes();
```
Comment (carterkozak) [inline okhttp-clients/src/main/java/com/palantir/conjure/java/okhttp/RemotingOkHttpCall.java:313]: Don't NPE when the body is null

File: okhttp-clients/src/main/java/com/palantir/conjure/java/okhttp/RemotingOkHttpCall.java [hunk 3]
```diff
@@ -342,14 +361,19 @@ private boolean shouldRetry(IOException exception, Optional<Duration> backoff) {
...
             Tags.AttemptSpan attemptSpan,
-            Runnable execution) {
+            Consumer<Callback> enqueueFunction,
+            Callback callback) {
         DetachedSpan backoffSpan = attemptSpan.attemptSpan().childDetachedSpan("OkHttp: backoff-with-jitter");
 
         // TODO(rfink): Investigate whether ignoring the ScheduledFuture is safe, #629.
         schedulingExecutor.schedule(
-                () -> executionExecutor.submit(() -> {
+                () -> executionExecutor.execute(() -> {
```
Comment (carterkozak) [inline okhttp-clients/src/main/java/com/palantir/conjure/java/okhttp/RemotingOkHttpCall.java:385]: if the provided runnable throws, execute will allow the uncaught exception handler to log it, while submit assumes the returned future is read (but we ignore it)

File: okhttp-clients/src/main/java/com/palantir/conjure/java/okhttp/RemotingOkHttpCall.java [hunk 4]
```diff
@@ -342,14 +361,19 @@ private boolean shouldRetry(IOException exception, Optional<Duration> backoff) {
...
                     backoffSpan.complete();
-                    execution.run();
+                    if (isCanceled()) {
+                        callback.onFailure(this, new SafeIoException("Request is cancelled"));
+                    } else {
+                        enqueueFunction.accept(callback);
+                    }
```
Comment (carterkozak) [inline okhttp-clients/src/main/java/com/palantir/conjure/java/okhttp/RemotingOkHttpCall.java:391]: If the calling thread is interrupted while waiting to retry, we don't want to send more requests.

Comment (carterkozak) [reply okhttp-clients/src/main/java/com/palantir/conjure/java/okhttp/RemotingOkHttpCall.java:391]: Alternatively we could track futures, but that gets gnarly and easy to break.

**Comment by CRogers** [comment]
This generally looks good to me, but for my own context was there anything that prompted these changes or just general making sure we're handling error/edge cases better?

**Comment by carterkozak** [comment]
I took a look over the code with an eye toward #1385. I don't expect this commit to solve that issue, but it may give us more information/narrow the search space.

**Comment by jkozlowski** [comment]
How testable are these improvements?

**Comment by CRogers** [comment]
+1 does seem a little strange to have these changes but no tests to demonstrate they work.

**Comment by carterkozak** [comment]
Added two tests that fail on develop, they're a bit frightening.

**Review by CRogers — APPROVED** [review]

**Pull request merged**